keras_spiking/model_energy.py

```diff
@@ -0,0 +1,810 @@
+    if any(x is None for x in output_shape):
+        raise ValueError(
+            "Cannot compute stats for '%s' layer because the output shape %s contains "
+            "undefined elements; if this represents a temporal axis, consider wrapping "
+            "this layer in ``tf.keras.layers.TimeDistributed``"
+            % (compat.node_layer(node).name, output_shape)
+        )
+    return np.prod(output_shape)
+
+
+def _act_stats(activation, n_neurons):
+    return (
+        {"neurons": 0, "spiking": True}
+        if activation in ("linear", tf.keras.activations.linear, None)
+        else {"neurons": n_neurons, "spiking": False}
+    )
+
+
+@ModelEnergy.register_layer(Layer)
+def layer_stats(node, **_):
+    """Fallback for computing stats on an unknown layer (assumes none)."""
+
+    # layer types that we know don't affect neurons/connections, so we don't register
+    # a stat computer for them
+    whitelist = (
+        Add,
+        Dropout,
+        Flatten,
+        GlobalAveragePooling1D,
+        GlobalAveragePooling2D,
+        GlobalAveragePooling3D,
+        InputLayer,
+        Reshape,
+    )
+
+    if not isinstance(compat.node_layer(node), whitelist):
+        warnings.warn(
+            "Cannot compute stats for layer of type %r."
+            "Use `ModelEnergy.register_layer` to register this layer."
+            % (type(compat.node_layer(node)).__name__,)
+        )
+    return {}
+
+
+@ModelEnergy.register_layer(Activation)
+@ModelEnergy.register_layer(ReLU)
+@ModelEnergy.register_layer(LeakyReLU)
+def act_stats(node):
+    """Compute activation layer stats."""
+    return _act_stats(
+        getattr(compat.node_layer(node), "activation", "relu"), _output_size(node)
+    )
+
+
+@ModelEnergy.register_layer(Conv1D)
+@ModelEnergy.register_layer(Conv2D)
+@ModelEnergy.register_layer(Conv3D)
+def conv_stats(node):
+    """Compute ``Conv1D``/``Conv2D``/``Conv3D`` layer stats."""
+
+    kernel_size = np.prod(compat.node_layer(node).kernel.shape[:-1])
+    output_size = _output_size(node)
+
+    result = {"connections": int(output_size * kernel_size)}
+    result.update(_act_stats(compat.node_layer(node).activation, output_size))
+    return result
+
+
+@ModelEnergy.register_layer(Dense)
+def dense_stats(node):
+    """Compute ``Dense`` layer stats."""
+
+    output_size = _output_size(node)
+    spatial_size = output_size // node.output_shapes[-1]
+    kernel_size = np.prod(compat.node_layer(node).kernel.shape)
+
+    result = {"connections": spatial_size * kernel_size}
+    result.update(_act_stats(compat.node_layer(node).activation, output_size))
+    return result
+
+
+@ModelEnergy.register_layer(SpikingActivation)
+def spikingactivation_stats(node):
+    """Compute `.SpikingActivation` layer stats."""
+    # note: using input shape because we know that will always have a temporal axis
+    # (whether return_sequences=true/false)
+    return {"neurons": np.prod(node.input_shapes[2:])}
```

Comment:
Shouldn't this have `"spiking": True`? Or am I not understanding this new spiking stat correctly?

Reply:
It could have `spiking=True`, but that's the default so it's not required.

Reply:
Ok, I see where that default happens now. I would explicitly return it as `True` here, just because we do it for other activations.